File: src/main/java/builder_java/Person.java
package builder_java;

/**
 * @author angus
 * @date 2019/11/20
 */
public class Person {
    private String name;
    private int age;
    private Contact contact;
    private Address addr;

    private Person(Builder builder) {
        this.name = builder.name;
        this.age = builder.age;
        this.contact = builder.contact;
        this.addr = builder.addr;
    }

    @Override
    public String toString() {
        return "Person{" +
                "name='" + name + '\'' +
                ", age=" + age +
                ", contact=" + contact +
                ", addr=" + addr +
                '}';
    }

    public static class Builder {
        private String name;
        private int age;
        private Contact contact;
        private Address addr;

        public Builder(String name, int age) {
            this.name = name;
            this.age = age;
        }

        public Builder contact(Contact contact) {
            this.contact = contact;
            return this;
        }

        public Builder addr(Address addr) {
            this.addr = addr;
            return this;
        }

        public Person build() {
            return new Person(this);
        }
    }

    public static void main(String[] args) {
        Person p = new Builder("Angus", 23)
                .addr(new Address("四川", "成都", "鹤山街道"))
                .contact(new Contact("18328635851", "1713862733", "Lu"))
                .build();
        System.out.println(p);
    }
}

File: src/main/kotlin/builder/a/Person.kt
package builder.a

/**
 * @author angus
 * @date 2019/11/21
 */
data class Person(
    var name: String? = null,
    var age: Int? = null,
    var contact: Contact? = null,
    var addr: Address? = null
)

data class Contact(
    var phone: String? = null,
    var qq: String? = null,
    var wechat: String? = null
)

data class Address(
    var province: String? = null,
    var city: String? = null,
    var street: String? = null
)

class Builder(init: Builder.() -> Unit) {
    var name: String? = null
    var age: Int? = null
    var contact: Contact? = null
    var addr: Address? = null

    init {
        init()
    }

    fun contact(init: Contact.() -> Unit) {
        contact = Contact().apply(init)
    }

    fun addr(init: Address.() -> Unit) {
        addr = Address().apply(init)
    }

    fun build(): Person = Person(name, age, contact, addr)
}

fun person(init: Builder.() -> Unit) = Builder(init).build()

fun main() {
    // kotlin 不需要 builder 模式？
    var p = Person(
        name = "Angus",
        age = 23,
        contact = Contact("18328635851", "1713862733", "Lu"),
        addr = Address("四川", "成都", "鹤山")
    )
    println(p)

    p = Person(name = "Angus", age = 23).apply {
        // 可以进行逻辑运算
        contact = Contact("18328635851", "1713862733", "Lu")
        addr = Address("四川").apply {
            city = "成都"
            street = "鹤山街道"
        }
    }
    println(p)

    // DSL Builder，DSL  像是下一代的建造者模式
    p = person {
        name = "Angus"
        age = 23
        contact {
            phone = "18328635851"
            qq = "1713862733"
            wechat = "Lu"
        }
        addr {
            province = "四川"
            city = "成都"
            street = "鹤山街道"
        }
    }
    println(p)
}